test/integration/revocation_test.go

```diff
@@ -32,109 +33,163 @@ func isPrecert(cert *x509.Certificate) bool {
 	return false
 }
 
-// TestPrecertificateRevocation tests that a precertificate without a matching
-// certificate can be revoked using all of the available RFC 8555 revocation
-// authentication mechanisms.
-func TestPrecertificateRevocation(t *testing.T) {
+// TestRevocation tests that a certificate can be revoked using all of the
+// RFC 8555 revocation authentication mechanisms. It does so for both certs and
+// precerts (with no corresponding final cert), and for both the Unspecified and
+// keyCompromise revocation reasons.
+func TestRevocation(t *testing.T) {
 	t.Parallel()
 	// Create a base account to use for revocation tests.
 	os.Setenv("DIRECTORY", "http://boulder:4001/directory")
-	c, err := makeClient("mailto:example@letsencrypt.org")
-	test.AssertNotError(t, err, "creating acme client")
 
-	// Create a specific key for CSRs so that it is possible to test revocation
-	// with the cert key.
-	certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
-	test.AssertNotError(t, err, "creating random cert key")
-
-	// Create a second account to test revocation with an equally authorized account
-	otherAccount, err := makeClient()
-	test.AssertNotError(t, err, "creating second acme client")
-	// Preauthorize a specific domain with the other account before it has been
-	// added to the ct-test-srv reject list.
-	preAuthDomain := random_domain()
-	_, err = authAndIssue(otherAccount, nil, []string{preAuthDomain})
-	test.AssertNotError(t, err, "preauthorizing second acme client")
-
-	testCases := []struct {
-		name         string
-		domain       string
-		revokeClient *client
-		revokeKey    crypto.Signer
-	}{
-		{
-			name:      "revocation by certificate key",
-			revokeKey: certKey,
-		},
-		{
-			name:      "revocation by owner account key",
-			revokeKey: c.Account.PrivateKey,
-		},
-		{
-			name:         "equivalently authorized account key",
-			revokeClient: otherAccount,
-			revokeKey:    otherAccount.Account.PrivateKey,
-			domain:       preAuthDomain,
-		},
+	type authMethod string
+	var (
+		byAccount authMethod = "byAccount"
+		byAuth    authMethod = "byAuth"
+		byKey     authMethod = "byKey"
+	)
+
+	type certKind string
+	var (
+		finalcert certKind = "cert"
+		precert   certKind = "precert"
+	)
+
+	type testCase struct {
+		method      authMethod
+		reason      int
+		kind        certKind
+		expectError bool
 	}
 
-	for _, tc := range testCases {
-		t.Run(tc.name, func(t *testing.T) {
-			// If the test case didn't specify a domain make one up randomly
-			if tc.domain == "" {
-				tc.domain = random_domain()
-			}
-			// If the test case didn't specify a different client to use for
-			// revocation use c.
-			if tc.revokeClient == nil {
-				tc.revokeClient = c
+	testCases := make([]testCase, 0)
+	for _, kind := range []certKind{precert, finalcert} {
+		for _, reason := range []int{ocsp.Unspecified, ocsp.KeyCompromise} {
+			for _, method := range []authMethod{byAccount, byAuth, byKey} {
```

Comment:
Nit, it would be a bit more idiomatic, and easier to check individual cases at a glance, if this test used a more table-driven style. Feel free to take or leave this though.

Reply:
This is a table driven test - it's just that the table is generated programmatically. Which is a good choice in this case because it ensures we cover the full combinatorial set of cases.

Reply:
Yeah I had the table all written out by hand for a little bit, but it's such a big table and so few things change in each entry that I collapsed it down to this and was personally much happier with it.